for_fedora/man/shinken-broker.8
.TH shinken-broker 8 "December 29, 2011" "Shinken User Manuals"
.SH NAME
.PP
shinken-broker - Shinken broker daemon
.SH SYNOPSIS
.PP
shinken-broker [-dr] [-c \f[I]CONFIGFILE\f[]] [\[em]debugfile
\f[I]DEBUGFILE\f[]]
.SH DESCRIPTION
.PP
Shinken broker daemon.
.PP
The shinken-broker's role is to export and manage data from schedulers
(such as status).
The management itself is done by modules.
.PP
The following management modules are included: * export into an NDO
(Nagios Data Out) database (MySQL or Oracle backend) * export to MERLIN
(Module for Effortless Redundancy and Loadbalancing In Nagios) database
(MySQL backend) * service-perfdata export * export to CouchDB
.PP
Multiple modules can be enabled simultaneously
.SH OPTIONS
.TP
.B --version
Show version number and exit
.RS
.RE
.TP
.B -c \f[I]CONFIGFILE\f[], --config \f[I]CONFIGFILE\f[]
Config file
.RS
.RE
.TP
.B -d, --daemon
Run in daemon mode
.RS
.RE
.TP
.B -r, --replace
Replace previously running broker
.RS
.RE
.TP
.B -h, --help
Print detailed help screen
.RS
.RE
.TP
.B --debugfile \f[I]DEBUGFILE\f[]
Enable debug logging to \f[I]DEBUGFILE\f[]
.RS
.RE
.SH AUTHORS
David Hannequin.
Arthur Gautier.
Michael Leinartas.


Changelog
########################
Shinken ChangeLog
#######################

1.4.2 - 06/03/2014
------------------
	CORE ENHANCEMENT (backport)
	Add : Triggers functions for perfdata
	Add : RST doc

	CORE FIXES (backport)
	Fix : Perfdata Parsing
	Fix : Arbiter parsing (Catch bad service_dependencies, timeperiod definitions)
	Fix : crash 'c.t_to_go is  None' in scheduler when time change
	Fix : adjust_worker_by_load was wrong, worker was not created again
	Fix : Hosts over-checked when services are critical (don't respect retry_interval)
	Fix : Prevent from returning next time notif in the past (don't over notify).

	MODULE ENHANCEMENT (backport)
	Add : Option to group collect plugin instances in one service
	Add : RST doc

	MODULES FIXES (backport)
	Fix : Graphite module skipping 0 values



1.4.1 - 02/01/2014
------------------

	CORE FIXES
	Fix : Protocol Error exception not defined
	Fix : Close FD on action timeout
	Fix : Plugins path in install script

	MODULES FIXES
	Fix : Webui (layout)
	Fix : Aws-Import-Arbiter error
																				

1.4   - 27/05/2013
------------------
	CORE ENHANCEMENT
	Add : Change Vsphere SDK Version
	Add : new option for realm : broker complete links, so now a scheduler can give broks to several brokers instead of only one.
	Add : now the arbiter can take its arbiter lookup name from parameter. Use to allow same host to run arbiter in HA for tests.
	Add : macro_modulations objects, to dynamically change the macro based on the timeperiod. Can be useful as check_modulations, but easier to use, so maybe they will simply replace them	Use persistent cookies by defaut
	Add : (nerocide) now KEY are expand on the service_dependencies	property.
	Add : first version of user/system time catch for checks.
	Add : check_modulations objects.	
	
	CORE FIXES
	Fix : (reported by:Georges Racinet) lot of SSL errors with Pyro3. Fix #839
	Fix : bad offline receiver was too much verbose. Fix #559	
	Fix : (reported by: Beuc) : bad discovery script name. Fix #843	
	FIX : #613 macros was overinding modules list instead of just adding a module. No bug at all just a misleading configuration in macro
	Fix : the mysql python module installation on redhat/centos
	FIX : Jenkins tests + autogeneration of all_tests based on py file in the tests directory
	FIX : Jenkis test scripts : The loop skip empty lines
	Fix : (reported by: miluebbe) catch bad protocol exception in daemons, like a telnet connexion. Fix #829
	Fix : Timeperiod calculation
	Fix : (reported by:darthelmet) bad NOT management for bp_rule and complex expression. Fix #827
	Fix : remove useless print.
	Fix : (reported by:sebguilbaud) bad URL for check_hpasm.
	Fix : try to fix the no log error for scheduler and log	level. Need a beter solution.
	Fix : Check if a poller exists to handle services tag.Raise error if not
	Fix : (reported by:boardstretcher) fix #814 about regression since last SSO related patch.
	Fix : get_all_host_names_set functions since 1.2.4.
	Fix : fix #509  now bp_rules can manage multi-levels rules with ()
	Fix : Missing quote in host state notification e-mails
	Fix : (reported by: Johan Svensson) now poller worker crashs are now put in the log.
	Fix : (reported by: flaf and Imrane.Dessai) bad additive inheritance with multiples templates.
	FIX : MySQL-python complainig about outdated version of distribute
	FIX : (by :Alexandre Veyrenc) graphite module now manage perfdata with space in names
	
	
	MODULES ENHANCEMENT
	Add : Sqlite to Mongodb logstore converter
	Add : ws_arbiter now supports multiple input values. The ws_arbiter module can now accept a list of metrics to process. See:http://www.shinken-monitoring.org/forum/index.php/topic,846.0.html
	Add : remote_user_variable from wsgi environment, to use set 'remote_user_enable 2'
	Add : collectd header in shinken-specific.cfg
	Add : proper support for counter/derive/absolute datatypes, by taking the derivative. Removed debug logging.
	Add : support for collectd messages. Added support for more than one value in VALUES collectd part. Fixed graphite broker to nicely deal with multiple values. Fixed graphite graphs to support	multiple values by sh
	Add : new arbiter module, FileTag, taht will tag an host if it is present in a flat file.	
	Add : support for high resolution time/interval (collectd 5.0),	and also added support for derive and absolute types
	Add : AWS arbiter module, to import your hosts from AWS/EC2.
	Add : (reported by: joachim schiele) update the PNP version to the lastest one.
	Ehn : add an optionnal list of hostnames to ignore for iptag module	
	Add : example client for zmq_broker.py	
	Add : A ZeroMQ pub socket broker for shinken
	
	MODULES FIXES
	Fix : (reported by:Benoit Dunand-Laisin) livestatus command management with json option.
	Fix : (reported by:pepejey) void password in ldap auth module. Fix #822	
	Fix : make the NDO synchronize_database_id option 1 by default to avoid single scheduler mode bug. Make the module far less verbose too.
	Fix : (reported by:Johan Svensson) nrpe booster module now manage more than 1024 simultaneous connexions by switchting from select based poll to a real poll call.
	Fix : mongodb fsync as a parameter
	Fix : invalid variable name causes nrpe_poller crash when recieving SSLError exception

src/shinken-broker.py
#!/usr/bin/env python
#Copyright (C) 2009-2010 : 
#    Gabes Jean, naparuba@gmail.com 
#    Gerhard Lausser, Gerhard.Lausser@consol.de
#
#This file is part of Shinken.
#
#Shinken is free software: you can redistribute it and/or modify
#it under the terms of the GNU Affero General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.
#
#Shinken is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU Affero General Public License for more details.
#
#You should have received a copy of the GNU Affero General Public License
#along with Shinken.  If not, see <http://www.gnu.org/licenses/>.


#This class is an interface for Broker
#The broker listen configuration from Arbiter in a port (first argument)
#the configuration gived by arbiter is schedulers where broker will take broks.
#When already launch and have a conf, broker still listen to arbiter
# (one a timeout)
#if arbiter whant it to have a new conf, broker forgot old chedulers
#(and broks into)
#take new ones and do the (new) job.

import time
import sys
import platform
import sys, os
import getopt


#We know that a Python 2.5 or Python3K will fail.
#We can say why and quit.
python_version = platform.python_version_tuple()

## Make sure people are using Python 2.5 or higher
if int(python_version[0]) == 2 and int(python_version[1]) < 6:
    print "Shinken require as a minimum Python 2.6.x, sorry"
    sys.exit(1)

if int(python_version[0]) == 3:
    print "Shinken is not yet compatible with Python3k, sorry"
    sys.exit(1)


try:
    import Pyro.core
except ImportError:
    print "Shinken require the Python Pyro module. Please install it."
    sys.exit(1)

#Module from 2.6 and higher
from Queue import Empty
from multiprocessing import active_children


from satellite import Satellite
from daemon import Daemon
from util import to_int, to_bool
from module import Module, Modules
from modulesmanager import ModulesManager
from log import Log
from brok import Brok

#Load to be used by modules
from resultmodulation import Resultmodulation
from escalation import Escalation
from timeperiod import Timeperiod
from contact import Contact
from command import Command, CommandCall

VERSION = "0.1"





#Interface for Arbiter, our big MASTER
#It put us our conf
class IForArbiter(Pyro.core.ObjBase):
	#We keep app link because we are just here for it
	def __init__(self, app):
		Pyro.core.ObjBase.__init__(self)
		self.app = app
		self.schedulers = app.schedulers
		self.arbiters = app.arbiters
		self.pollers = app.pollers
		self.reactionners = app.reactionners


	#function called by arbiter for giving us our conf
	#conf must be a dict with:
	#'schedulers' : schedulers dict (by id) with address and port
	#TODO: catch case where Arbiter send somethign we already have
	#(same id+add+port) -> just do nothing :)
	#REF : doc/shinken-conf-dispatching.png (4)
	def put_conf(self, conf):
		self.app.have_conf = True
		self.app.have_new_conf = True
		print "Sending us ", conf
		#If we've got something in the schedulers, we do not
		#want it anymore
		#self.schedulers.clear()
		for sched_id in conf['schedulers'] :
			s = conf['schedulers'][sched_id]
			self.schedulers[sched_id] = s
			uri = "PYROLOC://%s:%d/Broks" % (s['address'], s['port'])
			self.schedulers[sched_id]['uri'] = uri
			self.schedulers[sched_id]['broks'] = {}
			self.schedulers[sched_id]['instance_id'] = s['instance_id']
			self.schedulers[sched_id]['running_id'] = 0
			self.schedulers[sched_id]['active'] = s['active']
			#We cannot reinit connexions because this code in
			#in a thread, and
			#pyro do not allow thread to create new connexions...
			#So we do it just after.
		print "We have our schedulers :", self.schedulers

		#Now get arbiter
		for arb_id in conf['arbiters'] :
			a = conf['arbiters'][arb_id]
			self.arbiters[arb_id] = a
			uri = "PYROLOC://%s:%d/Broks" % (a['address'], a['port'])
			self.arbiters[arb_id]['uri'] = uri
			self.arbiters[arb_id]['broks'] = {}
			self.arbiters[arb_id]['instance_id'] = 0 #No use so all to 0
			self.arbiters[arb_id]['running_id'] = 0			
		print "We have our arbiters :", self.arbiters

		#Now for pollers
		for pol_id in conf['pollers'] :
			p = conf['pollers'][pol_id]
			self.pollers[pol_id] = p
			uri = "PYROLOC://%s:%d/Broks" % (p['address'], p['port'])
			self.pollers[pol_id]['uri'] = uri
			self.pollers[pol_id]['broks'] = {}
			self.pollers[pol_id]['instance_id'] = 0 #No use so all to 0
			self.pollers[pol_id]['running_id'] = 0			
		print "We have our pollers :", self.pollers
		
		#Now reactionners
		for rea_id in conf['reactionners'] :
                        r = conf['reactionners'][rea_id]
                        self.reactionners[rea_id] = r
                        uri = "PYROLOC://%s:%d/Broks" % (r['address'], r['port'])
                        self.reactionners[rea_id]['uri'] = uri
                        self.reactionners[rea_id]['broks'] = {}
                        self.reactionners[rea_id]['instance_id'] = 0 #No use so all to 0
                        self.reactionners[rea_id]['running_id'] = 0
                print "We have our reactionners :", self.reactionners
		
		if not self.app.have_modules:
			self.app.modules = conf['global']['modules']
			self.app.have_modules = True
			print "We received modules", self.app.modules
		

	#Arbiter ask us to do not manage a scheduler_id anymore
	#I do it and don't ask why
	def remove_from_conf(self, sched_id):
		try:
			del self.schedulers[sched_id]
		except KeyError:
			pass


	#Arbiter ask me which sched_id I manage, If it is not ok with it
	#It will ask me to remove one or more sched_id
	def what_i_managed(self):
		return self.schedulers.keys()


	#Use for arbiter to know if we are alive
	def ping(self):
		print "We ask us for a ping"
		return True


	#Use by the Arbiter to push broks to broker
	def push_broks(self, broks):
		self.app.add_broks_to_queue(broks.values())
		return True


	#Use by arbiter to know if we have a conf or not
	#can be usefull if we must do nothing but 
	#we are not because it can KILL US! 
	def have_conf(self):
		return self.app.have_conf


	#Call by arbiter if it thinks we are running but we must do not (like
	#if I was a spare that take a conf but the master returns, I must die
	#and wait a new conf)
	#Us : No please...
	#Arbiter : I don't care, hasta la vista baby!
	#Us : ... <- Nothing! We are die! you don't follow 
	#anything or what??
	def wait_new_conf(self):
		print "Arbiter want me to wait a new conf"
		self.schedulers.clear()
		self.app.have_conf = False



#Our main APP class
class Broker(Satellite):
	#default_port = 7772
	properties = {
		'workdir' : {'default' : '/usr/local/shinken/src/var', 'pythonize' : None, 'path' : True},
		'pidfile' : {'default' : '/usr/local/shinken/src/var/brokerd.pid', 'pythonize' : None, 'path' : True},
		'port' : {'default' : '7772', 'pythonize' : to_int},
		'host' : {'default' : '0.0.0.0', 'pythonize' : None},
		'user' : {'default' : 'shinken', 'pythonize' : None},
		'group' : {'default' : 'shinken', 'pythonize' : None},
		'idontcareaboutsecurity' : {'default' : '0', 'pythonize' : to_bool},
		'modulespath' : {'default' :'/usr/local/shinken/src/modules' , 'pythonize' : None, 'path' : True}
		}
	

	def __init__(self, config_file, is_daemon, do_replace, debug, debug_file):
		self.print_header()

		#From daemon to manage signal. Call self.manage_signal if
		#exists, a dummy function otherwise
		self.set_exit_handler()

                #Log init
		self.log = Log()
		self.log.load_obj(self)

		#The config reading part
		self.config_file = config_file
		#Read teh config file if exist
		#if not, default properties are used
		self.parse_config_file()

		if config_file != None:
		        #Some paths can be relatives. We must have a full path by taking
                        #the config file by reference
			self.relative_paths_to_full(os.path.dirname(config_file))

                #Check if another Scheduler is not running (with the same conf)
		self.check_parallele_run(do_replace)
                
                #If the admin don't care about security, I allow root running
		insane = not self.idontcareaboutsecurity

                #Try to change the user (not nt for the moment)
                #TODO: change user on nt
		if os.name != 'nt':
			self.change_user(insane)
		else:
			Log().log("Sorry, you can't change user on this system")

                #Now the daemon part if need
		if is_daemon:
			self.create_daemon(do_debug=debug, debug_file=debug_file)

		#Bool to know if we have received conf from arbiter
		self.have_conf = False
		self.have_new_conf = False
		#Ours schedulers
		self.schedulers = {}
		self.mod_instances = [] # for brokers from modules

		#Our arbiters
		self.arbiters = {}

		#Our pollers and reactionners
		self.pollers = {}
		self.reactionners = {}
		
		#Modules are load one time
		self.have_modules = False
		self.modules = []

		#All broks to manage
		self.broks = [] #broks to manage
		#broks raised this turn and that need to be put in self.broks
		self.broks_internal_raised = [] 



	#Manage signal function
	#TODO : manage more than just quit
	#Frame is just garbage
	def manage_signal(self, sig, frame):
		Log().log("\nExiting with signal %s" % sig)
		self.daemon.shutdown(True)
		sys.exit(0)


        #Schedulers have some queues. We can simplify call by adding
        #elements into the proper queue just by looking at their type
        #Brok -> self.broks
	#TODO : better tag ID?
	def add(self, elt):
		if isinstance(elt, Brok):
                        #For brok, we TAG brok with our instance_id
			elt.data['instance_id'] = 0
			self.broks_internal_raised.append(elt)
			return


	#Get teh good tabs for links by the kind. If unknown, return None
	def get_links_from_type(self, type):
		t = {'scheduler' : self.schedulers, 'arbiter' : self.arbiters, \
		     'poller' : self.pollers, 'reactionner' : self.reactionners}
		if type in t :
			return t[type]
		return None

	

	#initialise or re-initialise connexion with scheduler or
	#arbiter if type == arbiter
	def pynag_con_init(self, id, type='scheduler'):
		#Get teh good links tab for looping..
		links = self.get_links_from_type(type)
		if links == None:
			Log().log('DBG: Type unknown for connexion! %s' % type)
			return
		
		if type == 'scheduler':
                        #If sched is not active, I do not try to init
		        #it is just useless
			is_active = links[id]['active']
			if not is_active:
				return

		print "Init connexion with", links[id]['uri']
		running_id = links[id]['running_id']
		uri = links[id]['uri']
		links[id]['con'] = Pyro.core.getProxyForURI(uri)

		try:
			#intial ping must be quick
			links[id]['con']._setTimeout(5)
			links[id]['con'].ping()
			new_run_id = links[id]['con'].get_running_id()
			#data transfert can be longer
			links[id]['con']._setTimeout(120)

		        #The schedulers have been restart : it has a new run_id.
		        #So we clear all verifs, they are obsolete now.
			if links[id]['running_id'] != 0 or new_run_id != running_id:
				links[id]['broks'].clear()
			        #we must ask for a enw full broks if
			        #it's a scheduler
				if type == 'scheduler':
					print "Ask for a broks generation"
					links[id]['con'].fill_initial_broks()
			links[id]['running_id'] = new_run_id			
		except Pyro.errors.ProtocolError, exp:
			Log().log(str(exp))
			return
		except Pyro.errors.NamingError, exp:
			Log().log("%s is not initilised : %s" % (type, str(exp)))
			links[id]['con'] = None
			return
		except KeyError , exp:
			Log().log("%s is not initilised : %s" % (type, str(exp)))
                        links[id]['con'] = None
                        return

		Log().log("Connexion OK")


	#Get a brok. Our role is to put it in the modules
	#THEY MUST DO NOT CHANGE data of b !!!
	#REF: doc/broker-modules.png (4-5)
	def manage_brok(self, b):
		to_del = []
		#Call all modules if they catch the call
		for mod in self.modules_manager.get_internal_instances():
			try:
				mod.manage_brok(b)
			except Exception as exp:
				print exp.__dict__
				Log().log("Warning : The mod %s raise an exception: %s, I kill it" % (mod.get_name(),str(exp)))
				print "DBG:", type(exp)
				to_del.append(mod)
		#Now remove mod that raise an exception
		for mod in to_del:
			self.modules_manager.remove_instance(mod)

	
	#Add broks (a tab) to different queues for
	#internal and external modules
	def add_broks_to_queue(self, broks):
		#Ok now put in queue brocks for manage by
		#internal modules
		self.broks.extend(broks)
		
		#and for external queues
		#REF: doc/broker-modules.png (3)
		for b in broks:
			for q in self.modules_manager.get_external_to_queues():
				q.put(b)

				
	#Each turn we get all broks from
	#self.broks_internal_raised and we put them in
	#self.broks
	def interger_internal_broks(self):
		self.add_broks_to_queue(self.broks_internal_raised)
		self.broks_internal_raised = []


	#Get 'objects' from external modules
	#from now nobody use it, but it can be useful
	#for a moduel like livestatus to raise external
	#commandsfor example
	def get_objects_from_from_queues(self):
		for f in self.modules_manager.get_external_from_queues():
			full_queue = True
			while full_queue:
				try:
					o = f.get(block=False)
					self.add(o)
				except Empty :
					full_queue = False
				


	#We get new broks from schedulers
	#REF: doc/broker-modules.png (2)
	def get_new_broks(self, type='scheduler'):
		#Get teh good links tab for looping..
		links = self.get_links_from_type(type)
		if links == None:
			Log().log('DBG: Type unknown for connexion! %s' % type)
			return
		
		#We check for new check in each schedulers and put
		#the result in new_checks
		for sched_id in links:
			try:
				con = links[sched_id]['con']
				if con is not None: #None = not initilized
					tmp_broks = con.get_broks()
					for b in tmp_broks.values():
						b.instance_id = links[sched_id]['instance_id']

					#Ok, we can add theses broks to our queues
					self.add_broks_to_queue(tmp_broks.values())

				else: #no con? make the connexion
					self.pynag_con_init(sched_id, type=type)
                        #Ok, con is not know, so we create it
			except KeyError as exp: 
				#print exp
				self.pynag_con_init(sched_id, type=type)
			except Pyro.errors.ProtocolError as exp:
				Log().log(str(exp))
				#we reinitialise the ccnnexion to pynag
				self.pynag_con_init(sched_id, type=type)
                        #scheduler must not #be initialized
			except AttributeError as exp: 
				Log().log(str(exp))
                        #scheduler must not have checks
			except Pyro.errors.NamingError as exp:
				Log().log(str(exp))
			# What the F**k? We do not know what happenned,
			#so.. bye bye :)
			except Exception,x: 
				Log().log(''.join(Pyro.util.getPyroTraceback(x)))
				sys.exit(0)




	#modules can have process, and they can die
	def check_and_del_zombie_modules(self):
                #Active children make a join with every one, useful :)
		act = active_children()
		self.modules_manager.check_alive_instances()


	#Main function, will loop forever
	def main(self):

		Pyro.config.PYRO_STORAGE = self.workdir
		Pyro.config.PYRO_MULTITHREADED = 0
                #Daemon init
		Pyro.core.initServer()

		Log().log("Opening port: %s" % self.port)
		self.daemon = Pyro.core.Daemon(host=self.host, port=self.port)
		#If the port is not free, pyro take an other. I don't like that!
		if self.daemon.port != self.port:
			Log().log("Error : the port %d was not free" % self.port)
			sys.exit(1)
		self.uri2 = self.daemon.connect(IForArbiter(self),"ForArbiter")

                #We wait for initial conf
		self.wait_for_initial_conf()

		#Do the modules part, we have our modules in self.modules
		#REF: doc/broker-modules.png (1)
		self.modules_manager = ModulesManager('broker', self.modulespath, self.modules)
		self.modules_manager.load()
		self.mod_instances = self.modules_manager.get_instances()


                #Connexion init with Schedulers
		for sched_id in self.schedulers:
			self.pynag_con_init(sched_id, type='scheduler')
			
		for pol_id in self.pollers:
			self.pynag_con_init(pol_id, type='poller')

		for rea_id in self.reactionners:
                        self.pynag_con_init(rea_id, type='reactionner')


		#Now main loop
		i = 0
		#timeout = 1.0
		while True:
			i = i + 1
			if not i % 50:
				print "Loop ", i
			#begin_loop = time.time()

			#Begin to clean modules
			self.check_and_del_zombie_modules()

			#Now we check if arbiter speek to us in the daemon.
			#If so, we listen for it
			#When it push us conf, we reinit connexions
			self.watch_for_new_conf(0.0)
			
			#Maybe the last loop we raised some broks internally
			#we should interger them in broks
			self.interger_internal_broks()

			#Now we can get new broks from arbiters in self.broks
			#self.get_new_broks(type='arbiter')
			#And from schedulers
			self.get_new_broks(type='scheduler')
			#And for other satellites
			self.get_new_broks(type='poller')
			self.get_new_broks(type='reactionner')
			
			
		        #We must had new broks at the end of the list, so we reverse the list
			self.broks.reverse()			
			
			start = time.time()
			while(len(self.broks) != 0):
				now = time.time()
				#Do not 'manage' more than 1s, we must get new broks
				#every 1s
				if now - start > 1:
					break

				b = self.broks.pop()
			        #Ok, we can get the brok, and doing something with it
				#REF: doc/broker-modules.png (4-5)
				self.manage_brok(b)
				
				#Ok we manage brok, but we still want to listen to arbiter
				self.watch_for_new_conf(0.0)

		        #Restore the good sense
			self.broks.reverse()

			#Maybe external modules raised 'objets'
			#we should get them
			self.get_objects_from_from_queues()
			
			#Maybe we do not have something to do, so we wait a little
			if len(self.broks) == 0:
				self.watch_for_new_conf(1.0)

			#TODO : sleep better...
			#time.sleep(1)



################### Process launch part
def usage(name):
    print "Shinken Broker Daemon, version %s, from :" % VERSION
    print "        Gabes Jean, naparuba@gmail.com"
    print "        Gerhard Lausser, Gerhard.Lausser@consol.de"
    print "Usage: %s [options] [-c configfile]" % name
    print "Options:"
    print " -c, --config"
    print "\tConfig file."
    print " -d, --daemon"
    print "\tRun in daemon mode"
    print " -r, --replace"
    print "\tReplace previous running scheduler"
    print " -h, --help"
    print "\tPrint detailed help screen"
    print " --debug"
    print "\tDebug File. Default : no use (why debug a bug free program? :) )"


#lets go to the party
if __name__ == "__main__":
    #Manage the options
    try:
        opts, args = getopt.getopt(sys.argv[1:], "hrdc::w", ["help", "replace", "daemon", "config=", "debug=", "easter"])
    except getopt.GetoptError, err:
        # print help information and exit:
        print str(err) # will print something like "option -a not recognized"
        usage(sys.argv[0])
        sys.exit(2)
    #Default params
    config_file = None
    is_daemon=False
    do_replace=False
    debug=False
    debug_file=None
    for o, a in opts:
        if o in ("-h", "--help"):
            usage(sys.argv[0])
            sys.exit()
	elif o in ("-r", "--replace"):
            do_replace = True
        elif o in ("-c", "--config"):
            config_file = a
        elif o in ("-d", "--daemon"):
            is_daemon = True
	elif o in ("--debug"):
            debug = True
	    debug_file = a
        else:
            print "Sorry, the option",o, a, "is unknown"
	    usage(sys.argv[0])
            sys.exit()

    p = Broker(config_file, is_daemon, do_replace, debug, debug_file)
    #import cProfile
    p.main()
    #command = """p.main()"""
    #cProfile.runctx( command, globals(), locals(), filename="var/Shinken.profile" )


